### PR TITLE
Store IDL definitions in MCAP

### PR DESCRIPTION
**Pull request by james-rms**
Adds logic to detect IDL-defined message types and store the IDL.

### CODE REVIEW

File: rosbag2_storage_mcap/include/rosbag2_storage_mcap/message_definition_cache.hpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  MessageSpec(MessageDefinitionFormat format, std::string text, const std::string& package_context);
+  const std::set<std::string> dependencies;
+  const std::string text;
+  MessageDefinitionFormat format;
+};
+
+class MessageDefinitionCache final {
+public:
+  /**
+   * Concatenate the message definition with its dependencies into a self-contained schema.
```
Comment (james-rms) [inline rosbag2_storage_mcap/include/rosbag2_storage_mcap/message_definition_cache.hpp:62]: now out of date

File: rosbag2_storage_mcap/src/mcap_storage.cpp [hunk 2]
```diff
@@ -483,10 +483,14 @@ void MCAPStorage::write(std::shared_ptr<const rosbag2_storage::SerializedBagMess
   const auto& datatype = topic_info.topic_metadata.type;
   const auto schema_it = schema_ids.find(datatype);
   if (schema_it == schema_ids.end()) {
-    std::string full_text = msgdef_cache_.get_full_text(datatype);
+    auto [format, full_text] = msgdef_cache_.get_full_text(datatype);
     mcap::Schema schema;
     schema.name = datatype;
-    schema.encoding = "ros2msg";
+    if (format == rosbag2_storage_mcap::internal::MessageDefinitionFormat::ROS2MSG) {
+      schema.encoding = "ros2msg";
```
Comment (james-rms) [inline rosbag2_storage_mcap/src/mcap_storage.cpp]: leave link to spec appendix

**Comment by emersonknapp** [comment]
Can a `.msg` use an IDL-defined subtype? If so, does this handle that? I couldn't quite tell from the code

**Comment by james-rms** [comment]
You're right, it can, and this code doesn't handle that (it'll fail with "file not found"). I can take a look at falling back to IDL if any dependency has no MSG definition.

**Comment by james-rms** [comment]
@emersonknapp ok, handled and tested.

**Review by emersonknapp — APPROVED** [review]

**Pull request merged**